src/Grid/grid.jl

```diff
@@ -0,0 +1,242 @@
+# Grid types
+export Node, Cell, CellIndex, CellFace, FaceIndex, Grid
+
+# Cell type alias
+export Line, QuadraticLine,
+       Triangle, QuadraticTriangle, Quadrilateral, QuadraticQuadrilateral,
+       Tetrahedron, QuadraticTetrahedron, Hexahedron, QuadraticHexahedron
+
+# Grid utilities
+export getcells, getncells, getnodes, getnnodes, getcelltype,
+       getcellset,  getnodeset, getfaceset, getcoordinates, getcoordinates!,
+       getcellsets, getnodesets, getfacesets, onboundary, nfaces
+
+export addnodeset!, addcellset!, transform
+
+#########################
+# Main types for meshes #
+#########################
+"""
+A `Node` is a point in space.
+"""
+immutable Node{dim, T}
+    x::Vec{dim, T}
+end
+Node{dim, T}(x::NTuple{dim, T}) = Node(Vec{dim, T}(x))
+getcoordinates(n::Node) = n.x
+
+#=
+#A bit redundant but perhaps we want to put more in here:
+immutable Face
+    onboundary::Bool
+end
+onboundary(f::Face) = f.onboundary
+Face() = Face(false)
+=#
+
+"""
+A `Cell` is a sub-domain defined by a collection of `Node`s as it's vertices.
+"""
+immutable Cell{dim, N, M}
+    nodes::NTuple{N, Int}
+    onboundary::NTuple{M, Bool}
```

Comment:
It feels weird to store this here. Why not in a separate vector in the `grid`? We already have the cell id in the cell iterator so it is easy to retrieve anyway.